File: server/src/routes/anki.ts
import { Router } from 'express';
import * as ankiService from '../services/anki.js';
import type { SearchNotesRequest, CreateNoteRequest } from 'shared';

export const ankiRouter = Router();

// GET /api/anki/decks - List all deck names
ankiRouter.get('/decks', async (_req, res) => {
  try {
    const decks = await ankiService.getDecks();
    res.json({ decks });
  } catch (error) {
    console.error('Error fetching decks:', error);
    res.status(500).json({ error: 'Failed to fetch decks. Is Anki running?' });
  }
});

// GET /api/anki/models - List all note types
ankiRouter.get('/models', async (_req, res) => {
  try {
    const models = await ankiService.getModels();
    res.json({ models });
  } catch (error) {
    console.error('Error fetching models:', error);
    res.status(500).json({ error: 'Failed to fetch models. Is Anki running?' });
  }
});

// GET /api/anki/models/:name/fields - Get fields for a note type
ankiRouter.get('/models/:name/fields', async (req, res) => {
  try {
    const fields = await ankiService.getModelFields(req.params.name);
    res.json({ fields });
  } catch (error) {
    console.error('Error fetching model fields:', error);
    res.status(500).json({ error: 'Failed to fetch model fields' });
  }
});

// POST /api/anki/search - Search notes by query
ankiRouter.post('/search', async (req, res) => {
  try {
    const { query } = req.body as SearchNotesRequest;
    if (!query) {
      res.status(400).json({ error: 'Query is required' });
      return;
    }
    const notes = await ankiService.searchNotes(query);
    res.json({ notes });
  } catch (error) {
    console.error('Error searching notes:', error);
    res.status(500).json({ error: 'Failed to search notes' });
  }
});

// POST /api/anki/notes - Create a new note/card
ankiRouter.post('/notes', async (req, res) => {
  try {
    const { deckName, modelName, fields, tags } = req.body as CreateNoteRequest;

    if (!deckName || !modelName || !fields) {
      res.status(400).json({ error: 'deckName, modelName, and fields are required' });
      return;
    }

    const noteId = await ankiService.createCard({
      deck: deckName,
      model: modelName,
      word: fields.Word || '',
      definition: fields.Definition || '',
      exampleSentence: fields.Example || '',
      sentenceTranslation: fields.Translation || '',
      tags,
    });

    res.json({ noteId });
  } catch (error) {
    console.error('Error creating note:', error);
    res.status(500).json({ error: 'Failed to create note' });
  }
});

// GET /api/anki/notes/:id - Get note details
ankiRouter.get('/notes/:id', async (req, res) => {
  try {
    const noteId = parseInt(req.params.id, 10);
    if (isNaN(noteId)) {
      res.status(400).json({ error: 'Invalid note ID' });
      return;
    }

    const note = await ankiService.getNoteById(noteId);
    if (!note) {
      res.status(404).json({ error: 'Note not found' });
      return;
    }

    res.json({ note });
  } catch (error) {
    console.error('Error fetching note:', error);
    res.status(500).json({ error: 'Failed to fetch note' });
  }
});

// GET /api/anki/status - Check AnkiConnect status
ankiRouter.get('/status', async (_req, res) => {
  try {
    const connected = await ankiService.testConnection();
    res.json({ connected });
  } catch {
    res.json({ connected: false });
  }
});

File: python-server/anki_defs/routes/anki.py
"""Anki routes — proxy to AnkiConnect via httpx."""

from __future__ import annotations

import logging

import httpx
from bottle import Bottle, request, response

from ..services import ai, anki_connect

log = logging.getLogger(__name__)

_VALID_CARD_TYPES = {"vocab", "cloze", "mcCloze"}


def register(app: Bottle) -> None:
    @app.get("/api/anki/decks")
    def get_decks() -> dict:
        try:
            decks = anki_connect.get_decks()
            return {"decks": decks}
        except httpx.HTTPError as e:
            log.error("Error fetching decks: %s", e)
            response.status = 500
            return {"error": "Failed to fetch decks. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error fetching decks: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.get("/api/anki/models")
    def get_models() -> dict:
        try:
            models = anki_connect.get_models()
            return {"models": models}
        except httpx.HTTPError as e:
            log.error("Error fetching models: %s", e)
            response.status = 500
            return {"error": "Failed to fetch models. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error fetching models: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.get("/api/anki/models/<name>/fields")
    def get_model_fields(name: str) -> dict:
        try:
            fields = anki_connect.get_model_fields(name)
            return {"fields": fields}
        except httpx.HTTPError as e:
            log.error("Error fetching model fields: %s", e)
            response.status = 500
            return {"error": "Failed to fetch model fields. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error fetching model fields: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.post("/api/anki/search")
    def search_notes() -> dict:
        body = request.json or {}
        query = body.get("query", "")
        if not query:
            response.status = 400
            return {"error": "Query is required"}
        try:
            notes = anki_connect.search_notes(query)
            return {"notes": notes}
        except httpx.HTTPError as e:
            log.error("Error searching notes: %s", e)
            response.status = 500
            return {"error": "Failed to search notes. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error searching notes: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.post("/api/anki/notes")
    def create_note() -> dict:
        body = request.json or {}
        deck = body.get("deck", "")
        card_type = body.get("cardType", "vocab")
        word = body.get("word", "")

        if not deck:
            response.status = 400
            return {"error": "deck is required"}
        if card_type not in _VALID_CARD_TYPES:
            response.status = 400
            return {"error": f"Invalid cardType: {card_type}"}
        if not word and card_type == "vocab":
            response.status = 400
            return {"error": "word is required for vocab cards"}

        try:
            if not body.get("approveMigration"):
                pending = anki_connect.check_migrations_for_deck(deck, [card_type])
                if pending:
                    response.status = 409
                    return {"migrationRequired": True, "migrations": pending}

            note_id, model_name = anki_connect.create_card(
                deck=deck,
                card_type=card_type,
                word=word,
                definition=body.get("definition", ""),
                native_definition=body.get("nativeDefinition", ""),
                example=body.get("example", ""),
                translation=body.get("translation", ""),
                vocab_templates=body.get("vocabTemplates"),
                tags=body.get("tags"),
            )
            return {"noteId": note_id, "modelName": model_name}
        except httpx.HTTPError as e:
            log.error("Error creating note: %s", e)
            response.status = 503
            return {"error": "Could not connect to Anki. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error creating note: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.get("/api/anki/notes/<note_id:int>")
    def get_note(note_id: int) -> dict:
        try:
            note = anki_connect.get_note_by_id(note_id)
            if note is None:
                response.status = 404
                return {"error": "Note not found"}
            return {"note": note}
        except httpx.HTTPError as e:
            log.error("Error fetching note: %s", e)
            response.status = 503
            return {"error": "Could not connect to Anki. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error fetching note: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.delete("/api/anki/notes/<note_id:int>")
    def delete_note(note_id: int) -> dict:
        try:
            note = anki_connect.get_note_by_id(note_id)
            if note is None:
                response.status = 404
                return {"error": "Note not found"}
            if "auto-generated" not in note.get("tags", []):
                response.status = 403
                return {"error": "Cannot delete notes not created by this app"}
            anki_connect.delete_note(note_id)
            return {"success": True}
        except httpx.HTTPError as e:
            log.error("Error deleting note: %s", e)
            response.status = 503
            return {"error": "Could not connect to Anki. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error deleting note: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.post("/api/anki/sync")
    def sync_anki() -> dict:
        try:
            anki_connect.sync()
            return {"success": True}
        except httpx.HTTPError as e:
            log.error("Error syncing Anki: %s", e)
            response.status = 500
            return {"error": "Failed to sync. Is Anki running?"}
        except RuntimeError as e:
            log.error("Error syncing Anki: %s", e)
            response.status = 500
            return {"error": str(e)}

    @app.get("/api/anki/status")
    def get_status() -> dict:
        try:
            connected = anki_connect.test_connection()
            return {"connected": connected}
        except (httpx.HTTPError, RuntimeError):
            return {"connected": False}

    @app.get("/api/anki/languages")
    def get_languages() -> dict:
        return {"languages": ai.get_available_languages()}
